Add support for the OpenTitan chip

beezow opened the pull request:
This PR includes a OpenTitan Demo App which utilizes the AON Timer and the hubris test suite for OpenTitan.
It also includes a change to `qemu.sh` to support larger images.

This will not run on upstream qemu, but with Tylers updated version. https://gitlab.ba.rivosinc.com/rv/sw/ext/qemu/-/tree/dev/tkng/opentitan.

beezow committed:
chips: add `opentitan-earlgrey`

Co-authored-by: Tyler Ng <tkng@rivosinc.com>

beezow committed:
drv: added `opentitan-aontimer`.

This driver provides an interface to manage the opentitan AON/watchdog

Co-authored-by: Tyler Ng <tkng@rivosinc.com>

beezow committed:
task: added `aontimer`.

This task keeps the opentitan watchdog fed.

Co-authored-by: Tyler Ng <tkng@rivosinc.com>

beezow committed:
app: demo: add `demo-opentitan-earlgrey`.

This demo utilizes the opentitan AON.

Co-authored-by: Tyler Ng <tkng@rivosinc.com>

beezow committed:
chips: qemu: change device loader to use `final.ihex`

This change is needed as previously `final.bin` was used, but qemu
limits `.bin` image size to the ram size.  Switching to `.ihex` allows
the images to be larger than the ram in the vm.

beezow committed:
tests: add test suite to opentitan target

beezow committed:
Cargo fmt

beezow committed:
ci: add `demo-opentitan-earlgrey`

tkng-rivos commented on chips/opentitan/chip.toml:
```diff
@@ -0,0 +1,29 @@
+
+[mtimecmp]
+address = 0x4010_0118
+size = 0
```
The address of `mtime` and `mtimecmp` in QEMU right now uses the old addresses of the Ibex timer. Here's the old addresses:
- mtime: `0x4010_0104`
- mtimecmp: `0x4010_10C`

QEMU uses commit 217a0168ba118503c166a9587819e3811eeb0c0c for its OT implementation.

In the future when QEMU updates its devices, the addresses we have now will be correct.

beezow replied:
So this port is designed to run with your changes in [dev/tkng/opentitan](https://gitlab.ba.rivosinc.com/rv/sw/ext/qemu/-/tree/dev/tkng/opentitan).  I went ahead and [rebased](https://gitlab.ba.rivosinc.com/rv/sw/ext/qemu/-/tree/dev/drew/opentitan) them onto rivos/main the other day.

This branch has the correct addresses, right?

tkng-rivos replied:
If correct means the latest addresses in the Ibex timer spec, then yes.

tkng-rivos commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,82 @@
+
+    int_ctrl.enable_int(BARK_INT).unwrap();
+
+    // TODO this callback doesn't actually get called.
```
Update?

beezow replied:
Was actually just a logging problem.

tkng-rivos commented on app/demo-opentitan/src/main.rs:
```diff
@@ -0,0 +1,36 @@
+    };
+
+    //In RISC-V, this is just what mtimecmp gets set to.
+    const MTIMECMP: u32 = 1000;
```
This might be too low for the actual HW, but going from the current QEMU args (timebase-freq of 1,000,000) it's fine.

beezow replied:
Good point, I added a comment to reflect this.

tkng-rivos commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,82 @@
+
+// NOTE: you will probably want to remove this when you write your actual code;
+// we need to import userlib to get this to compile, but it throws a warning
+// because we're not actually using it yet!
```
Comment likely not needed anymore

beezow replied:
yep, removed the import but missed the comment.

beezow commented:
Looks like the CI didn't actually pass, just failed to evaluate.  @FawazTirmizi Can we make that fail the check?

furquan-rivos commented on chips/opentitan/chip.toml:
```diff
@@ -0,0 +1,29 @@
+[gpio0]
```
Are these peripheral addresses and interrupts taken from some specific version of the datasheet/specification? If so, can you please add a reference to it in the commit message and/or this file?

beezow replied:
They are taken from https://github.com/qemu/qemu/blob/master/hw/riscv/opentitan.c#L32.  Will add a comment.

furquan-rivos commented on build/xtask/src/flash.rs:
```diff
@@ -164,7 +164,7 @@ pub fn config(
         | "gimletlet-2" | "gimlet-a" | "gimlet-b" | "psc-1" | "sidecar-1"
         | "stm32g031" | "stm32g070" | "stm32g0b1" | "hifive-inventor"
-        | "hifive1-revb" => {
+        | "hifive1-revb" | "opentitan" => {
```
Is the chip named opentitan or is it opentitan earlgrey? (https://docs.opentitan.org/hw/top_earlgrey/doc/)

beezow replied:
Hm, it looks like it is "OpenTitan Earl Grey".  I would say that we should use the proper chip name, but also most the time it has just been referred to as the opentitan in our own docs.  Qemu also calls it the opentitan; however, the qemu naming conventions for the whole system is a bit suspect (it incorrectly calls all of the opentitan peripherals `ibex`).

Looking at the other chip names for baseline, I think Ill just change it to `opentitan_earlgrey`, unless someone has objections.

furquan-rivos replied:
From [tock documentation](https://github.com/tock/tock/blob/master/boards/README.md#out-of-tree-boards), it looks like `opentitan` is the project name and `earlgrey` is the chip/MCU name.

`opentitan_earlgrey` here sounds good to me.

furquan-rivos commented on drv/opentitan-aontimer/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+semihosting-riscv = [ "riscv-semihosting", "userlib/log-semihosting" ]
+log-null = ["userlib/log-null"]
+
+[target.'cfg(target_arch = "riscv32")'.dependencies]
```
Wouldn't this be always true?

beezow replied:
For this hardware peripheral we know it will only ever be on riscv, but hypothetically you could have an equivalent behaving arm peripheral.  The driver I think is otherwise arch agnostic and if complied for arm, these crates are not needed.
This is all a stretch though, just throwing ideas out there.  Alternatively, we could add a compile time check to prevent this driver from being compiled outside riscv, then just always include them.

furquan-rivos replied:
I think we can leave it in here. I was just curious what strategy we want to follow for writing the `Cargo.toml` files. We might see a lot of crates that probably would be used for RV only.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+
+#[allow(unused_imports)]
+use userlib::*;
+// use zerocopy::AsBytes;
```
Remove commented out code

beezow replied:
done

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+
+use core::ptr::slice_from_raw_parts_mut;
+
+const WDOG_REGWEN: usize = 4;
```
These should all be `u32`. They are fixed width 32-bit as per specification.

beezow replied:
Fixed.

beezow replied:
Actually, these are just indexes into an array, not the registers themselves.  Will adjust the name to reflect this as it also confused me.

furquan-rivos replied:
Ack

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+        unsafe {
+            //1: init regs
+            (*inst.base_addr)[WDOG_COUNT] = 0;
+            (*inst.base_addr)[INTR_STATE] = 0b11;
```
Can you please define these register fields instead of using magic bits?

beezow replied:
WIll change.  This does raise the question if we should use libraries such as[ tock-registers](https://crates.io/crates/tock-registers).  Probably not needed here, but we should decide if we recommend it or not as people start developing more tasks.  I know @FawazTirmizi has done some research on this.  I think there were code size concerns?

FawazTirmizi replied:
Yeah, there is some concern about code size, although we haven't done much proper analysis. For now I think just defining consts or an enum for the values is enough.

furquan-rivos replied:
After reading the code, I had the same question in my mind. If we should bring in the tock-register libraries for cleaner register field access. @FawazTirmizi - do you have the code size concerns captured somewhere? If not, can you please start a doc or bug to start tracking this activity?

beezow replied:
I just created some `const`.  Don't think it warranted adding tock.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+            return false;
+        }
+        unsafe {
+            (*self.base_addr)[WDOG_CTRL] |= 1;
```
Define register fields instead of using 1 and 0.

beezow replied:
> WIll change. This does raise the question if we should use libraries such as[ tock-registers](https://crates.io/crates/tock-registers). Probably not needed here, but we should decide if we recommend it or not as people start developing more tasks. I know @FawazTirmizi has done some research on this. I think there were code size concerns?

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+
+    // Converts  milliseconds to an appropriate count in the register.
+    // If the result is greater than max int, the value is saturated to max int.
+    fn ms_to_reg(&self, ms: u64) -> u32 {
```
nit: should this be named ms_to_count?

tkng-rivos replied:
Probably would be clearer. Or `ms_to_wdog_count`/`ms_to_reg_count` would also work.

beezow replied:
Went with `ms_to_reg_count`.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+// immediately reset.
+pub struct AonTimer {
+    base_addr: *mut [u32],
+    clock_freq: u32,
```
Use suffix _hz in the name so that it is clear that the clock frequency is stored in Hz?

beezow replied:
Done.

furquan-rivos commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,75 @@
+
+    int_ctrl.disable_int(BARK_INT).unwrap();
+
+    const AONTIMER_BASE: u32 = 0x4047_0000;
```
AON timer base address is fixed for the opentitan chip. Why does this need to be supplied in the task as well? Shouldn't the driver have this available from chip.toml?

beezow replied:
This feature is in the queue, https://rivosinc.atlassian.net/browse/SW-431.  Looking upstream, it seems Oxide has always hard coded their peripheral addresses.

furquan-rivos replied:
Thanks for the pointer, Drew. Yeah, it seems weird that the address is being duplicated in chip.toml and also in the tasks. I see that upstream oxide uses some chip-level crate to pull in the addresses in the drivers which helps avoid the duplicate address in the tasks directly. 

My other point here was that I would expect the driver to be aware of the base address but not the task directly i.e. the hierarchy will be something like:
some task that wants to use timer -----> driver for the timer ----> interrupt controller

So, this way the tasks don't deal directly with internals of the timer. Instead, they rely on the driver and the driver knows about the base address and also deals with the interrupt controller.

beezow replied:
I linked this thread into the jira issue, and also added a TODO in the code.

furquan-rivos commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,75 @@
+    int_ctrl.disable_int(BARK_INT).unwrap();
+
+    const AONTIMER_BASE: u32 = 0x4047_0000;
+    const AONTIMER_FREQ: u32 = 200_000;
```
Is this frequency configurable by each task? Or is this something that is fixed for the open titan chip?

beezow replied:
@rivos-tkng Would know for sure, but I think this constant is used to configure the AON timer.  So it would be configurable by whatever task owns it.

beezow replied:
As a slightly related side question,  @FawazTirmizi, can two task have access to the same peripheral in the chip.toml?

FawazTirmizi replied:
Yeah, it should be. I haven't tested it yet, but some of Oxide's apps like the [gimlet-rot](https://github.com/rivosinc/hubris/blob/2563c2c4550c0f0a558364aea200ddacc227ca2a/app/gimlet-rot/app.toml) do this.

beezow replied:
Hmm, that could cause problems with two task both trying to manage one device.  It may be an interesting feature if you could define a memory region to be exclusive to one task.  Then if two task both `use` it, it could provide a compile time error.

furquan-rivos replied:
I think having multiple tasks own a single peripheral isn't going to work without a lot of runtime coordination between the two. I think the way this will have to be organized is that there is a single peripheral driver and any task that wants to use the peripheral talks to the driver to get its work done.

tkng-rivos replied:
The timer frequency is not configurable from SW. This was supposed to be a C-style macro for converting time -> register count.

beezow replied:
If the frequency is fixed, it sounds like the type of thing that should be in the chip.toml

beezow replied:
Since it is fixed by the hardware, it seems like it should be in the chip.toml.  However, I don't know how hard it is to add arbitrary fields.  @FawazTirmizi

beezow replied:
I went ahead and move the timer frequency to the app.toml along with the bark/bite intervals.

furquan-rivos commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,75 @@
+const AONTIMER_BARK: u32 = 0x1;
+const AONTIMER_TIME: u32 = 0x2;
+
+const BARK_INT: u32 = 0;
```
Isn't this interrupt # fixed for the chip? If so, shouldn't this be provided by the driver/chip?

beezow replied:
So the interrupt number here is actually just an index into the interrupts available to the task.  

However, looking at the [original hubris docs](https://hubris.oxide.computer/reference/#_interrupts_from_a_tasks_perspective)
 "When calling irq_control, the task names its interrupts by the notification bits they will set when they fire, rather than their hardware interrupt numbers" 
With this in mind, I think the call to `int_ctrl.enable_int` should probably use the notification bit, not just an index into its interrupts.  This would stay more inline with how interrupts are normally handled.  @FawazTirmizi do you have some context on this?

This change would also make the `BARK_INT` and `AONTIMER_BARK` constants the same, and the shared value would not have to be hard-coded after this [ticket](https://rivosinc.atlassian.net/browse/SW-530) is done.

furquan-rivos replied:
Yes, I don't think `BARK_INT` and `AONTIMER_BARK` need to be two separate constants. Both really seem to be referring to the same thing i.e. notification for bark interrupt.

beezow replied:
These have been consolidated, but will still need to wait for that ticket to remove the magic numbers.

furquan-rivos commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,75 @@
+
+task_slot!(INT_CONTROLLER, ext_int_ctrl);
+
+const AONTIMER_BARK: u32 = 0x1;
+const AONTIMER_TIME: u32 = 0x2;
```
Are these notification bits posted by the plic driver? If so, then these should be suffixed with _NOTIFICATION to make it clear. Also, how does the plic know that it has to set bit 1 i.e. value 0x2? I don't see that being added to app.toml

BTW, looking at the way this gets used, it seems very brittle and redundant. app.toml needs to define the exact same notification bits which are defined here so that the task gets notified of the interrupt correctly. I think there should be just one place where this information is defined.

beezow replied:
Good idea with the `_NOTIFICATION` suffix.  

Within `app/demo-opentitan/app.toml`: 
```
[tasks.ext_int_ctrl.config]
ints = [158]
tasks = ['"aontimer"']
notification = [1]
base = 0x4800_0000
pbits = 3
```
Each of `ints`, `tasks`, `notification`, acts as a list where all of the elements of the same index are associated.  So this block defines that the `aontimer` task has access to interrupt `158`, using notification bit `1`.

Then in `task/aontimer/src/main.rs`, the call to `sys_set_timer` defines the notification bit to use:
```
sys_set_timer(Some(alarm + 1000), AONTIMER_TIME_NOTIFICATION);
```

The ability to extract the notification bit for the app is being tracked by this jira [ticket](https://rivosinc.atlassian.net/browse/SW-530).  Also see this [comment](https://github.com/rivosinc/hubris/pull/7#discussion_r972465239) on the initial plic PR.  I think this feature was held out for now because of some extra complications that would be needed in the build system, but @FawazTirmizi would have more context.

beezow replied:
After talking to @FawazTirmizi, it seems this may have exposed some bugs with the PLIC, but still have to look into it.

furquan-rivos replied:
With respect to using notification bit 2, I just posted a comment [here](https://github.com/rivosinc/hubris/pull/12/commits/620556c9b24c63c4c0dc0dc53e926f2802cd01d3#r974698623). It looks like the AON timer device supports 2 kinds of interrupts since there are 2 timers within the timer device - watchdog timer (bark) and wakeup timer (wakeup). It looks like the driver implementation adds support for watchdog timer only, but this app is making use of both interrupts. That was my source of confusion as to how AONTIMER_TIME_NOTIFICATION was being set.

> After talking to @FawazTirmizi, it seems this may have exposed some bugs with the PLIC, but still have to look into it.

What's the bug in PLIC?

beezow replied:
Bug fixed. https://github.com/rivosinc/hubris/pull/27

furquan-rivos commented on chips/fe310-g002/qemu.sh:
```diff
@@ -1 +1 @@
-qemu-system-riscv32 -M sifive_u,msel=1 -nographic -serial mon:stdio -device loader,addr=0x20010000,cpu-num=0,file=final.bin -semihosting -semihosting-config enable=on,userspace=on
+qemu-system-riscv32 -M sifive_u,msel=1 -nographic -serial mon:stdio -device loader,addr=0x20010000,cpu-num=0,file=final.ihex -semihosting -semihosting-config enable=on,userspace=on
```
Unable to post a comment on the commit so posting it here. Why does qemu allow `.ihex` but not '.bin`? Does it mean that the image is being loaded over something that is not RAM? Is it because this is being fetched from flash?

beezow replied:
It does allow `.bin`. However, when checking if the image is within the allowable size, the max size for a `.bin` is the size of the ram (in fact it only does this check for binaries).  See the[ qemu generic loader](https://github.com/qemu/qemu/blob/b3cd3b5a66f0dddfe3d5ba2bef13cd4f5b89cde9/hw/core/generic-loader.c#L158).  I think ihex contains addresses, so qemu can extract the size, but it cannot do so with a raw binary(?)

furquan-rivos replied:
I think `.bin` is an ELF so it should have the required information about where different components end up in RAM. I am just worried if using `.ihex` is working around some incorrect address loading being caught by the `.bin`/elf path. Did you trace the elf path to see what information it is looking for and based on what it decides that the required regions of the binary cannot fit in RAM?

beezow replied:
I did trace the path for the `.bin` but it just goes straight to the size check, even before trying to parse the data.  
While on the topic though, I do have an elf bug I still need to hunt down.  There seems to be an address issue with the final elf.  When trying to use the final.elf with qemu, it is getting incorrectly loaded.

furquan-rivos replied:
> I did trace the path for the .bin but it just goes straight to the size check, even before trying to parse the data.

I think we should make the loading of elf and ihex both work. Can you please raise an issue for this? We can debug that separately. I don't think we need to block this PR for that.

> While on the topic though, I do have an elf bug I still need to hunt down. There seems to be an address issue with the final elf. When trying to use the final.elf with qemu, it is getting incorrectly loaded.

Interesting. Let's raise bugs for each of these issues and track them.

beezow replied:
The `.elf` should be fixable, but I think loading the `.bin` would require a qemu patch.  As long as the `.bin` is smaller than the ram size, it loads and runs just fine.

https://rivosinc.atlassian.net/browse/SW-555
and 
https://rivosinc.atlassian.net/browse/SW-554

I am not sure if the `.bin` problem belongs in the hubris epic, but we can always move it.

beezow commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+            (*inst.base_addr)[WDOG_COUNT] = 0;
+            (*inst.base_addr)[INTR_STATE] = 0b11;
+            (*inst.base_addr)[WDOG_BARK_THOLD] = inst.ms_to_reg(bark_thold_ms);
+            if bark_thold_ms > bite_thold_ms {
+                (*inst.base_addr)[WDOG_BITE_THOLD] =
+                    inst.ms_to_reg(bark_thold_ms);
+            } else {
+                (*inst.base_addr)[WDOG_BITE_THOLD] =
+                    inst.ms_to_reg(bite_thold_ms);
```
Should we impose this on the user without it being clear that `bite_thold_ms` must be greater than `bark_thold_ms`?

beezow replied:
Added compile time check

beezow commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,150 @@
+    // If the result is greater than max int, the value is saturated to max int.
+    fn ms_to_reg(&self, ms: u64) -> u32 {
+        let mut time: u64 = ms * ((self.clock_freq / 1000) as u64);
+        if time > u32::MAX.into() {
+            time = u32::MAX.into();
+        }
+        time.try_into().unwrap()
```
Is silently truncating to `u32` the right behavior here?  Should we return a result so the caller knows the desired `ms` was too big?

furquan-rivos replied:
I think returning error would be better since silently assuming a value would mean that the caller doesn't realize that the timer expires sooner than intended. I am hoping we don't ever see that long of a timeout but might make sense to return appropriate error.

beezow replied:
Changed.

FawazTirmizi reviewed (changes requested):
`drv-ot-aontimer` isn't really a driver, it's just a library for interfacing with the timer. This isn't something that should be exposed for any task to use. The library should be made a part of the `aontimer` task, which should in turn be moved into the `drv` folder.

furquan-rivos commented:
> drv-ot-aontimer isn't really a driver, it's just a library for interfacing with the timer. This isn't something that should be exposed for any task to use. The library should be made a part of the aontimer task, which should in turn be moved into the drv folder.

That is what I was thinking as well. The organization doesn't look correct. I was thinking this would be something like this:

drv-timer-api : API for interfacing with the timer driver. This is what will be used by a task

drv-timer : Timer driver (what is currently split into two drv and task)

task : This would be using the drv-timer-api to configure timers and callbacks.

furquan-rivos commented:
This PR seems to have more commits than probably what you intended. I see 6 commits from Fawaz which are not related. Something wrong with the rebasing?

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,149 @@
+    // If the bark threshold > bite threshold, the bite threshold is set to
+    // the bark threshold.
+    pub fn new(
+        base_addr: u32,
```
The way the code is currently organized where this driver is a library, probably having base address as an input is fine. But, the base address isn't really something that is task configurable. It is fixed for a chip. So, is there any benefit of allowing base address to be taken in as an input?

tkng-rivos replied:
You're right. At the time it was written, it was meant to be a more generic interface for watchdogs, but that changed as time went on.

beezow replied:
I was actually thinking the opposite.  With the driver being in its own crate, I think it should be agnostic to the chip memory map.  So the task using the driver will tell the crate what addresses to operate at.

furquan-rivos replied:
So, it depends on what the driver is intended for. Let's say the driver is meant only for opentitan earlgrey chip. Then, passing in a base address is unnecessary. However, if the driver is meant to be used by (let's say for example) multiple opentitan chips with different memory layouts, then yes, accepting base address as an input here makes sense. However, in either case, I don't think the address should be coming in from a "task". Task should not have to worry about any chip-level details.

beezow replied:
Once there are[ build consts for peripherals](https://rivosinc.atlassian.net/browse/SW-431), the task can just pass the const though to the driver crate.  It might also be possible for the driver crate to directly access the const.

furquan-rivos replied:
We can keep this as is for now. But, we should revisit this in the future. I think accepting this address from tasks also opens up the condition that task might pass in an incorrect address and so needs to be handled correctly. Instead if it is a chip attribute, it should just be something that is embedded directly within the driver.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,149 @@
+            bark_cb,
+        };
+        unsafe {
+            //1: init regs
```
What's `1:`?

beezow replied:
@rivos-tkng

tkng-rivos replied:
Comment no longer needed

tkng-rivos replied:
Was supposed to be a commented list, i.e. 1) init regs 2) do other stuff as notes when code was being written

beezow commented:
> > drv-ot-aontimer isn't really a driver, it's just a library for interfacing with the timer. This isn't something that should be exposed for any task to use. The library should be made a part of the aontimer task, which should in turn be moved into the drv folder.
> 
> That is what I was thinking as well. The organization doesn't look correct. I was thinking this would be something like this:
> 
> drv-timer-api : API for interfacing with the timer driver. This is what will be used by a task
> 
> drv-timer : Timer driver (what is currently split into two drv and task)
> 
> task : This would be using the drv-timer-api to configure timers and callbacks.

This is the ideal organization for drivers.  This demo is a little awkward in that it does not expose a driver for other task.  Looking at the original [hubris doc](https://hubris.oxide.computer/reference/#_when_not_to_use_a_server).  What is currently called `drv-ot-timer` should be a crate.  The [Drivers](https://hubris.oxide.computer/reference/#drivers) section also has some good context.  After reading this, I think the current organization is not far off.

I renamed the `task-aontimer` to `task-aontimer-demo` to better reflect what it is.  I also changed the import name for the `drv-ot-aontimer` to be `drv-ot-aontimer` rather than just `aontimer`

beezow commented:
> This PR seems to have more commits than probably what you intended. I see 6 commits from Fawaz which are not related. Something wrong with the rebasing?

Yea, this pr was originally made when the plic was still in PR, so it has  changed underneath it. Ill sort it out.

beezow commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    //Disable the watchdog timer. If the configuration is locked, returns false.
+    pub fn disable(&self) -> bool {
```
Should all of these functions be returning `Result` rather than `bool`?  The return value is to indicate success, which I think a `Result` is more indicative of.

tkng-rivos replied:
Probably. I'm not sure if it's even necessary to return some feedback on whether the configuration is locked, though it could be useful when testing. The locking is enforced within HW too.

beezow commented on task/aontimer/src/main.rs:
```diff
@@ -0,0 +1,79 @@
+    int_ctrl.enable_int(BARK_INT).unwrap();
+
+    let timer = aon::AonTimer::new(
+        AONTIMER_BASE,
+        AONTIMER_FREQ,
+        2000,
+        5000,
+        Some(|| {
```
I noticed other apps have configs within the app.toml. I think we should do the same here.
```
[tasks.aontimer_demo.config]
clock_frequency = ...
bark_threshold = ...
bite_threshold = ...
```
Then we could ensure the bark < bite relationship at compile time.

beezow replied:
Changed

furquan-rivos commented on chips/opentitan-earlgrey/chip.toml:
```diff
@@ -0,0 +1,32 @@
+[aontimer]
+address = 0x4047_0000
+size = 0x1000
+interrupts = { bark = 158 }
```
Why is this named differently (`bark`) than other interrupts that use `irq`? I understand that the aontimer spec refers to this threshold as bark, but it eventually triggers an interrupt. Do we want a different name here?

---

Also, this `bark` element doesn't really get used anywhere. I would expect something in app.toml to use aontimer.bark and support in build/xtask to check this `bark` key. But, neither exists. So, wondering if this is just a leftover?

tkng-rivos replied:
I don't know if `irq = ...` is a convention, if it is a convention it should've been `irq = 158` instead. It used to be used in the opentitan task toml, but there might have been changes that make this no longer necessary. @beezow

furquan-rivos replied:
Looking at [how xtask handles this](https://github.com/rivosinc/hubris/blob/2563c2c4550c0f0a558364aea200ddacc227ca2a/build/xtask/src/dist.rs#L1725), it seems to support any name for the interrupt. So, basically app.toml should be able to use something like `aontimer.bark`. Given that, I think it should be okay to use `bark` here.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,141 @@
+            (*inst.base_addr)[INTR_STATE_IDX] = 0b11;
+            (*inst.base_addr)[WDOG_BARK_THOLD_IDX] = inst.ms_to_reg_count(bark_thold_ms)?;
+            if bark_thold_ms > bite_thold_ms {
+                return Err(());
```
Should the error be more descriptive as to what failed?

beezow replied:
Fair point.  I would like to be able to move this to a compile time error though.
> I noticed other apps have configs within the app.toml. I think we should do the same here.
> 
> ```
> [tasks.aontimer_demo.config]
> clock_frequency = ...
> bark_threshold = ...
> bite_threshold = ...
> ```
> 
> Then we could ensure the bark < bite relationship at compile time.

furquan-rivos replied:
Makes sense.

beezow replied:
Done

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,141 @@
+}
+
+impl AonTimer {
+    // Initialize a new watchdog driver device with the specified base address,
```
Reading the [AON timer spec](https://docs.opentitan.org/hw/ip/aon_timer/doc/), the AON timer device supports 2 timers - watchdog timer and wakeup timer. It looks like this patch adds support for watchdog timer, but not the wakeup timer? However, the following patch for demo actually makes use of both the watchdog timer as well as the wakeup timer. Seems like there is a disconnect?

furquan-rivos replied:
I think I figured out my source of confusion here. As mentioned above there are two timers within the AON timer device. The demo app is configuring and using both timers. But, the wakeup timer is handled by the kernel in hubris, whereas the watchdog timer is handled by the AON timer driver.

furquan-rivos commented:
> > > drv-ot-aontimer isn't really a driver, it's just a library for interfacing with the timer. This isn't something that should be exposed for any task to use. The library should be made a part of the aontimer task, which should in turn be moved into the drv folder.
> > 
> > 
> > That is what I was thinking as well. The organization doesn't look correct. I was thinking this would be something like this:
> > drv-timer-api : API for interfacing with the timer driver. This is what will be used by a task
> > drv-timer : Timer driver (what is currently split into two drv and task)
> > task : This would be using the drv-timer-api to configure timers and callbacks.
> 
> This is the ideal organization for drivers. This demo is a little awkward in that it does not expose a driver for other task. Looking at the original [hubris doc](https://hubris.oxide.computer/reference/#_when_not_to_use_a_server). What is currently called `drv-ot-timer` should be a crate. The [Drivers](https://hubris.oxide.computer/reference/#drivers) section also has some good context. After reading this, I think the current organization is not far off.
> 
> I renamed the `task-aontimer` to `task-aontimer-demo` to better reflect what it is. I also changed the import name for the `drv-ot-aontimer` to be `drv-ot-aontimer` rather than just `aontimer`

Understood about the `task-aontimer-demo`. But, I think long term we will have to think if the watchdog driver (`drv-ot-aontimer`) should be a library that gets built with some task or should it be a separate task that gets built as a binary crate. In the latter case, you are also going to need idolatory definition and API library that will be instead integrated into the app crate.

furquan-rivos commented on app/demo-opentitan-earlgrey/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+version = "0.1.0"
+
+[features]
+#semihosting = ["kern/klog-semihosting-riscv"]
```
I think this should be dropped as there is no klog support in upstream sys/kern anymore.

beezow replied:
Done.

furquan-rivos commented on app/demo-opentitan-earlgrey/src/main.rs:
```diff
@@ -0,0 +1,37 @@
+    };
+
+    // TODO this will need to change when run on actual hardware
+    const MTIMECMP: u32 = 1000;
```
Why is this named `MTIMECMP`? This argument to start_kernel is supposed to be `tick_divisor`.

tkng-rivos replied:
From what I understand, this is what the `mtimecmp` register is set to, and the kernel periodicially resets `mtime` back to 0.

furquan-rivos replied:
I think that is an internal implementation detail of the rv32 architecture ;). The function signature expects `tick_divisor` i.e. the divisor to convert machine ticks into kernel ticks. It just seems a little weird to use MTIMECMP directly here.

beezow committed:
WIP change aontimer stuff to config rather than hardcode

beezow committed:
rebased onto main

beezow committed:
Add compile time check for bark/bite threshold

beezow committed:
opentitan demo: removed old comment from Cargo.toml

beezow committed:
drv: opentitan-aontimer: remove magic numbers

beezow commented:
So there seems to have been a regression with `mtime` working, so this isn't ready quite yet.

beezow committed:
cargo fmt

beezow committed:
add license to demo task

beezow committed:
fixup ci

beezow commented on task/aontimer-demo/build.rs:
```diff
@@ -0,0 +1,12 @@
+task_config::task_config!{
```
added

beezow commented:
This should be ready to go.  I tested both `demo-opentitan-earlgrey` and `tests-opentitan-earlgrey` in qemu.
I am going to leave the commit history for now to help with review, but will clean it up before merging.

furquan-rivos commented on app/demo-opentitan-earlgrey/src/main.rs:
```diff
@@ -0,0 +1,37 @@
+    // The opentitan requires enabling the timer and interrupts
+    // for mtime and mtimecmp to behave as expected
+    // See https://docs.opentitan.org/hw/ip/rv_timer/doc/
+    const OT_TIMER_CTRL: u32 = 0x4010_0004;
```
Shouldn't these be provided as services by some driver or kernel instead of having the task do this on its own by accessing the raw addresses?

FawazTirmizi replied:
The base address for the timer can be retrieved now through `task_peripherals_str()` in `build-util`.

beezow replied:
I think they could eventually get replaced to a call of something like `start_opentitan_timer()` but we currently don't have an opentitan crate to put this in.  
The app main seems like the best place to put something that needs to run before the kernel.  Where in a task do you think it should go?

furquan-rivos replied:
> I think they could eventually get replaced to a call of something like start_opentitan_timer() but we currently don't have an opentitan crate to put this in.

That sounds good to me.

> The app main seems like the best place to put something that needs to run before the kernel. Where in a task do you think it should go?

I missed the fact that this is run as part of the kernel before the `start_kernel` call. Makes sense to keep it here. Like you mentioned above once we have an opentitan crate this can end up there.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,154 @@
+            (*inst.base_addr)[WDOG_COUNT_IDX] = 0;
+            // write 1 to clear any pending interrupts
+            (*inst.base_addr)[INTR_STATE_IDX] =
+                INTR_STATE_BITS_WKUP | INTR_STATE_BITS_WDOG;
```
This whole driver is dealing with watchdog only. So, why does it enable the wakeup timer interrupt?

tkng-rivos replied:
So according to the [spec](https://docs.opentitan.org/hw/ip/aon_timer/doc/#Reg_intr_state), the code here is clearing the `INTR_STATE` register by writing 1 to each bit.

furquan-rivos replied:
Right. I meant that this should be `(*inst.base_addr)[INTR_STATE_IDX] = INTR_STATE_BITS_WDOG;` instead since `INTR_STATE_BITS_WKUP` is not really related to watchdog timer.

tkng-rivos replied:
If it's been decided already to have the driver solely focus on the watchdog part of the AON Timer IP, then I would go as far as to say to remove any reference to the wkup timer from this driver, since it's not needed and raises confusion.

furquan-rivos replied:
> If it's been decided already to have the driver solely focus on the watchdog part of the AON Timer IP, then I would go as far as to say to remove any reference to the wkup timer from this driver, since it's not needed and raises confusion.

That was exactly the motivation behind my question :) - to understand if this driver is supposed to deal with "watchdog" only or both "watchdog" and "wakeup". There are just 3 places where we reference wakeup timer here -
1. INTR_STATE_BITS_WKUP definition
2. clear_wdt_irq
3. new

And the comment on line 35 says "Initialize a new watchdog driver device". I am fine either ways. Just wanted to understand the plan here.

tkng-rivos replied:
Originally when it was written it was supposed to be both, but because we didn't know exactly what we were going to do with the wakeup timer (as it wasn't implemented yet and we have no way of testing it at the moment), it was punted to the todo-list. 

The demo app itself is concerned with the watchdog timer, but the driver could have both in the future, or we could have two separate drivers for each timer; their functions aren't related to each other. I'm not aware of any decision on that yet and I would imagine we'd want to wait until we know more about how the wakeup timer will work (or is implemented).

furquan-rivos replied:
In that case, would it make sense to just enable watchdog timer here and punt the addition of wakeup timer to later?

beezow replied:
Changed to just clear the watchdog interrupt bit.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    //Sets the bark threshold in ms. If the configuration is locked, returns false.
+    pub fn set_bark_thold(&self, bark_thold_ms: u64) -> Result<(), ()> {
```
Just noting that there are no checks in `set_bark_threshold` and `set_bite_threshold` to ensure that bark and bite thresholds are not configured incorrectly w.r.t. each other. Do we need to expose these APIs at all?

beezow replied:
That is a fair point, although it is a valid configuration as far as the AON timer is concerned.  I think the check within the task that uses the driver is good enough.  
I will remove the `pub` for now.

furquan-rivos commented on drv/opentitan-aontimer/src/lib.rs:
```diff
@@ -0,0 +1,154 @@
+    pub fn clear_wdt_irq(&self) {
+        unsafe {
+            (*self.base_addr)[INTR_STATE_IDX] =
+                INTR_STATE_BITS_WKUP | INTR_STATE_BITS_WDOG;
```
Name suggestes that watchdog interrupt is being cleared, but the code actually clears both watchdog and wakeup interrupts. Why is this touching the wakeup timer bit?

beezow replied:
changed to just the watchdog.

tkng-rivos commented on task/aontimer-demo/src/main.rs:
```diff
@@ -0,0 +1,83 @@
+        .unwrap();
+        int_ctrl.complete_int(AONTIMER_BARK_NOTIFICATION).unwrap();
+        if result.operation & AONTIMER_BARK_NOTIFICATION != 0x0 {
+            sys_log!("{}", sys_get_timer().now);
```
Unless we still need these logs for debugging, I think these should be commented out or removed

beezow replied:
The whole point of the task is a demo, so I think it is fine to leave the prints in.

beezow committed:
Switch to calling functions for init of the AON timer and make bark/bite
threshold functions private.

beezow commented:
old
